src/Todo/App.purs

```diff
@@ -53,127 +53,188 @@ localStorageKey = "todomvc-purescript-state"
-  -- after every modification
-  , receiveProps: \a -> receiveProps $ a { setState = \f -> a.setStateThen f saveState }
-  , render:       \a -> render $ a { setState = \f -> a.setStateThen f saveState }
-  }
+data Action
+  = EditNewTodo (Maybe String)
+  | KbdNewTodo
+  | TaskCheck Int
+  | TaskUpdate Int String
+  | TaskDelete Int
+  | ClearCompleted
+  | CheckAllTasks (Maybe Boolean)
+  | UpdateVisibility Visibility
+  | LoadState State
+  | Noop
+
+component :: React.Component Props
+component = React.createComponent "App"
+
+app :: JSX
+app = React.make component
+  { initialState
+  , didMount
+  , render
+  , update
+  , didUpdate
+  } {}
   where
     -- This is the only place we can run stuff only at the first mount
-    receiveProps { state, setState, isFirstMount } = when isFirstMount do
+    didMount self@{ state } = do
+      let setVisibility visibility =
+            React.send self (UpdateVisibility visibility)
       -- On first mount, we start the navigation:
       -- we have something super simple here, in which we match on
       -- the hash string and execute a side effect.
       -- For something fancier we might want to use a parser.
       let matchRoutes hash = case hash of
-            "#/"          -> setState _ { visibility = All }
-            "#/active"    -> setState _ { visibility = Active }
-            "#/completed" -> setState _ { visibility = Completed }
+            "#/"          -> setVisibility All
+            "#/active"    -> setVisibility Active
+            "#/completed" -> setVisibility Completed
             otherwise     -> pure unit
       runEffectFn1 startNavigation matchRoutes
 
       -- Then we try to read if we had some state persisted in LocalStorage
       -- If yes, we overwrite the state with it
       persisted <- LocalStorage.getItem localStorageKey
-      setState \_ -> case persisted of
-        Just (oldState :: State) -> oldState
-        _                        -> state
+      case persisted of
+        Just (oldState :: State) -> React.send self (LoadState oldState)
+        _ -> pure unit
+
+    didUpdate self _ = do
+      saveState self.state
+
+    update self action =
+      case action of
+        EditNewTodo val ->
+          Update self.state { newTodo = fromMaybe "" val}
+
+        KbdNewTodo ->
+          let
+            newDescription = String.trim self.state.newTodo
+            hasNewDescription = not (String.null newDescription)
+            newTodo =
+              { description: newDescription
+              , id: self.state.uid
+              , completed: false
+              }
+          in
+           if hasNewDescription then
+             Update self.state
+               { newTodo = ""
+               , tasks = Array.cons newTodo self.state.tasks
+               , uid = self.state.uid + 1
+               }
+           else
+             NoUpdate
+
+        TaskCheck id ->
+          let
+            negateCheck task =
+              if task.id == id then
```

Comment:
Let's try to keep formatting consistent: I like how you align `if`, `then`, `else` in the next branch below, so we could do the same here